code/datums/action.dm

```diff
@@ -43,10 +43,27 @@
 				return
 			Remove(owner)
 		owner = M
+
+		//button id generation
+		var/counter = 0
```

Comment:
You are still doing it wrong.
you cannot use a numerical counter of any kind.
The actions list is not static, it changes mid round, and if an action ability is ever added or removed, the id will break.

You need a static, between-round, unchanging ID.

Reply:
Actual id (that is used to store screen location of button) is composed of that numerical part and action name. Numerical part is here for cases when you have multiple similar actions (like having multiple flashlights or something, each of which can be toggled via action). And the way numerical part is generated (it seems) guarantees that if two actions with same actionnames won't get same numerical part, and as result actual string id is also differs.

Even while there's no need for static, between-round, unchanging ID (because that list in prefs is not saved between rounds (at least I suppose it is not)), it is actually unchanging in terms that if you have only one action with that name, it will always have equal id of "[name]_1". The only cases id can change is when you have multiple similar actions, and in such cases the order you get them defines their ids.

Reply:
Problem with previous attempt at generation of ids was that if you have two items with similar action, A and B, equip A then B, then unequip A and equip it again, they both get same id.
With updated variant id generation is a bit more complicated, but it seems to work right in any case.

Reply:
so you're not saving it between rounds?
Literally what is the point then.

Reply:
Saving it for a course of round.
Some actions are regained very often (like mentioned mech actions), and being able to arrange them only once per round is kinda nice.
I don't know if it's good to save this data between rounds, because setups of actionbuttons for different roles and equipment can overlap, and it can be very confusing (like if you have similar button placement (in terms of positions) for runic golem actionbuttons and for mechs, and you probably want to keep them as they are if you're runic golem without a mech or human mech operator, but if you become runic golem mech operator, they'll most probably overlap with each other and you might want to change arrangement), while saving it only for a course of round most probably wont get you in such situations.

Reply:
https://github.com/tgstation/tgstation/blob/master/code/__HELPERS/unsorted.dm#L1471

You can use the GUID function to generate unique ID's

Reply:
This one looks like very randomized, so it's not suitable.

Reply:
It can be used to give every action it's own unique ID, yet it's still not good because you lose flexibility (you'll need to rearrange buttons every time you change action source(exosuit mech, internals tank, light source etc).
Current generator (one I wrote) allows different action sources of identical actions to understand that they're same and should share setup.

Reply:
yeah I wasn't thinking clearly when I suggested it, ignore me